web/src/stream_list.ts

```diff
@@ -1310,3 +1310,42 @@ export function get_current_stream_li(): JQuery | undefined {
 
     return $stream_li;
 }
+
+export function get_sorted_channel_ids_for_next_unread_navigation(
+    current_channel_id?: number,
+): number[] {
+    // Here is the priority order for the channels:
+    // 1. Channels in the current section if uncollapsed.
+    // 2. Channels in the uncollapsed section.
+    // 3. Channels in the collapsed section.
+
+    let current_section_id: string | undefined;
+    if (current_channel_id !== undefined) {
+        current_section_id = stream_list_sort.current_section_id_for_stream(current_channel_id);
+    }
+    // Get sorted section ids.
+    const sections = stream_list_sort.get_current_sections().map((section) => ({
+        id: section.id,
+        channels: section.streams,
+        is_collapsed: collapsed_sections.has(section.id),
+    }));
+
+    sections.sort((a, b) => {
+        if (!a.is_collapsed && a.id === current_section_id) {
+            return -1; // a is the current section and uncollapsed
+        }
+        if (!b.is_collapsed && b.id === current_section_id) {
+            return 1; // b is the current section and uncollapsed
+        }
+        if (a.is_collapsed && !b.is_collapsed) {
+            return 1; // b is uncollapsed, a is collapsed
+        }
+        if (!a.is_collapsed && b.is_collapsed) {
+            return -1; // a is uncollapsed, b is collapsed
+        }
+        // Keep the original order for sections with the same collapsed state
+        return 0;
```

Comment:
Often this can be done more readably by just having a scoring function that return 2 * (a.is_collapsed && a.id === current_section_id) + 1 * a.is_collapsed, and doing `return score(a) - score(b)`.

Reply:
Thanks! Updated!